crates/extensions/c8y_mapper_ext/src/operations/handler.rs

```diff
@@ -138,23 +141,70 @@ impl OperationHandler {
             message,
         };
 
-        let topic: Arc<str> = message.message.topic.name.clone().into();
+        let topic = Arc::from(message.message.topic.name.as_str());
 
-        let running_operation = self.running_operations.remove(&topic);
+        let status = match GenericCommandState::from_command_message(&message.message) {
+            Ok(command) if command.status.is_empty() && is_clearing_message(&message.message) => {
+                None
+            }
+            Ok(command) => Some(command.status),
+            Err(err) => {
+                error!(%err, ?message, "could not parse command payload");
+                return;
+            }
+        };
 
-        let running_operation =
-            running_operation.unwrap_or_else(|| RunningOperation::spawn(Arc::clone(&self.context)));
+        let current_operation = self.running_operations.entry(topic);
 
-        let operation_status = running_operation
-            .update(message)
-            .await
-            .expect("operation task should not panic");
+        match current_operation {
+            Entry::Vacant(entry) => {
+                let Some(status) = status else {
+                    debug!("unexpected clearing message");
+                    return;
+                };
+
+                let context = Arc::clone(&self.context);
+                let handle = tokio::spawn(async move { context.update(message).await });
 
-        match operation_status {
-            OperationStatus::Ongoing(operation) => {
-                self.running_operations.insert(topic, operation);
+                let running_operation = RunningOperation { handle, status };
+
+                entry.insert(running_operation);
+            }
+
+            Entry::Occupied(entry) => {
+                if status.as_ref().is_some_and(|s| *s == entry.get().status) {
+                    debug!(
+                        "already handling operation message with this topic and status, ignoring"
+                    );
+                    return;
+                }
+
+                let (key, operation) = entry.remove_entry();
+
+                // TODO(marcel):
+                //- what happens when clearing message is posted by other client while we're
+                // handling a message? E.g. if we're doing a download to handle SUCCESSFUL status, do
+                // we abort or keep going?
+                // - what if we get a message that moves back status, i.e. SUCCESSFUL -> EXECUTING?
+
+                // if handling clearing message, just join task and remove
+                let Some(status) = status else {
+                    operation
+                        .handle
+                        .await
+                        .expect("operation task should not panic");
```

Comment:
Although not done in this PR, I'm having second thoughts about that `expect` call. Isn't it safer to just log the operation handler failure, if at all one happens, rather than crashing the whole mapper?

Reply:
`context.update()` only panics when it can't send a message to the MQTT publisher via a channel. The MQTT sender it uses is a cloned from C8yMapper, so if this happens we can't send any MQTT messages. This is an unrecoverable failure, so it should be a panic. The actor runtime should restart the entrire actor.

The fact that we only panic on MQTT channel failure should become more clear after it's extracted.